nvflare/fuel/utils/config_factory.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import pathlib
+from typing import List, Optional
+
+from nvflare.fuel.utils.config import Config, ConfigFormat, ConfigLoader
+from nvflare.fuel.utils.json_config_loader import JsonConfigLoader
+from nvflare.fuel.utils.omegaconf_loader import OmegaConfLoader
+from nvflare.fuel.utils.pyhocon_loader import PyhoconLoader
+
+
+class ConfigFactory:
+    logger = logging.getLogger(__qualname__)
+
+    @classmethod
+    def ordered_search_extensions(cls) -> List:
+        search_sequence: List[ConfigFormat] = ConfigFormat.search_order()
+        extensions = []
+        for fmt in search_sequence:
+            exts = ConfigFormat.get_extensions(fmt.name)
+            if exts:
+                for ext in exts:
+                    extensions.append((fmt, ext))
+
+        return extensions
+
+    @staticmethod
+    def search_config_format(init_file_path, search_dirs: Optional[List[str]] = None) \
+            -> (Optional[ConfigFormat], Optional[str]):
+        """
+        find the configuration format and the location (file_path) for given initial init_file_path and search directories.
+        for example, the initial config file path given is config_client.json
+        the search function will ignore the .json extension and search "config_client.xxx" in the given directory in
+        specified extension search order. The first found file_path will be used as configuration.
+        the ".xxx" is one of the extensions defined in the configuration format.
+
+        Args:
+            init_file_path: initial file_path for the configuration
+            search_dirs: search directory. If none, the parent directory of init_file_path will be used as search dir
+
+        Returns: Tuple of None,None or ConfigFormat and real configuration path
+
+        """
+        logger = ConfigFactory.logger
+        if not search_dirs:
+            parent_dir = pathlib.Path(init_file_path).parent
+            search_dirs = [str(parent_dir)]
+        # we ignore the original extension
+        file_basename = os.path.splitext(pathlib.Path(init_file_path).name)[0]
+        for search_dir in search_dirs:
+            logger.debug(f"search file:{file_basename} basename, search dirs = {search_dirs}")
+            for root, dirs, files in os.walk(search_dir):
```

Comment:
Should not walk thru all sub folders, otherwise a wrong file could be found. Instead, only look at the files directly in the dir.
Also the extension sequence is not right - see my comment before about non-default should be searched first before searching default cross ext types.